### ISSUE TITLE
Test for Breaking Schema Changes in Travis/Test flow

### ISSUE THREAD

**Issue by vinnyjth** [ApollosProject/apollos-prototype#744]
### User Story

As a developer, my local schema should be run against the production schema to determine if there are any breaking changes. 
 
### Description
After our 1.0 release, we will not tolerate breaking changes to our core schema. We might allow them in the future against any "2.0" work, but we should have tools to detect breaking changes no matter what.

**A breaking change is defined as a change that would cause errors to occur on old versions of the app pointed at current versions of the API** This is normally changes in the name of fields, changes in input types, and deletions of fields/queries. 

### Acceptance Criteria

* [ ] Developers should be able to run the tool locally to see if there are any breaking changes.
* [ ] Travis.ci should report any breaking changes before merging a PR 

### Other Resources

* Idea for implementing https://www.apollographql.com/docs/platform/schema-validation#service-check-on-ci
* Another idea https://github.com/entria/graphql-find-breaking-changes-cli

**Comment by vinnyjth** [comment]
@redreceipt This might be a fun one for ya!

**Comment by redreceipt** [comment]
Love this. Also, I was reading about the `eslint-plugin-graphql` plugin that lints your client queries against a `.graphql` schema file. Could be really awesome too!

**Comment by vinnyjth** [comment]
Oh that would be pretty sweet! Feel free to add that as a part of this PR.

**Comment by redreceipt** [comment]
*TL;DR* regarding the two implementations above:

1. **Apollo Engine Validation**. Requires trace reporting (query analytics) which is just dropping an Engine API key in our server config. Would register schema with CI check and compare against the last 100 (or whatever) *actual* query calls in the wild. Comments in your PR with a link to a thorough diff report:

### Apollo Service Check

🔄 Validated your local schema against schema tag `staging` on service `engine`.
🔢 Compared **18 schema changes** against **100 operations** seen over the **last 24 hours**.
❌ Found **7 breaking changes** that would affect **3 operations** across **2 clients**

🔗 [View your service check details](https://engine.apollographql.com/service/engine/checks?...).

+ Community supported
+ Thorough with resolution details
+ Determines the real effect on current users
- Potentially wouldn't catch *every* breaking change
- Requires Apollo Engine registration

2. **`graphql-findbreakingchanges`**. Simple package dependency that compares old schema to new schema. For example, useful for comparing any changes to a 1.0 release. Will just add `graphql-findbreakingchanges oldSchema.graphql newSchema.graphql` as a line in Travis.

+ Simplicity
+ Will presumably catch *all* breaking changes
- Will need to generate and maintain `.graphql` files

I tend to like option 2 better for the simplicity of it. Unless generating these executable files is a lot more complex than I realize.

Another huge advantage of having a local schema file is we'll be able to lint our client-side queries against it, major win. I just can't figure out how to generate these files, there's a major knowledge gap I'm missing because everyone seems to gloss over where these `.graphql` files are...

**Comment by vinnyjth** [comment]
@redreceipt 

Regarding option 1, something to keep in mind is that we already have an Apollo Engine account. 

Regarding option 2, I agree with you that I like this option better. One thing to consider though is "where does `oldschema.graphql`" come from? I like this approach, but answering that question can get tricky. 

Let's say you keep it in `git`. If you push an update to the schema, it passes in Travis, the PR is merged,how do you make the `newSchema` become the `oldSchema`? You can push a commit, but then we have to unblock `master` :P 

Something we have done before, which would recommend, is making a network request to the production API and downloading the current schema API in production, and using that as `oldSchema`.

**Comment by redreceipt** [comment]
I think you would just leave `oldSchema` as whatever you want, maybe on a 1.0 release or a 2.0 release, and then generate the `newSchema` in Travis every PR. So like this:

```
- yarn generate-schema
- yarn findBreakingChanges generatedSchema.graphql templateSchema.graphql
```

`templateSchema.graphql` is the only one that lives in the repo and it never changes.

**Comment by vinnyjth** [comment]
What if you changed a field in `generatedSchema` and then changed the field again before you update `templateSchema.graphql`? Then if anybody updates to the first change, and then updates to the second change, they will have introduced a breaking change :/

**Comment by vinnyjth** [comment]
@redreceipt I don't hate that idea though at all, I like the idea that we could making breaking changes between major releases. However, I think we should make a habit of requiring every single push to master to be "non breaking"

**Comment by redreceipt** [comment]
I think I follow. Definitely gets tricky. We would almost need another release branch to house the `templateSchema` and then work in `master` could be breaking but you couldn't push to `release` without updating the template.

**Comment by redreceipt** [comment]
*...or* you just turn off the check on a release push to master so it would be your only opportunity to introduce breaking changes

**Comment by redreceipt** [comment]
The problem with never allow breaking changes to master means a massive schema with a neverending list of deprecations

**Comment by redreceipt** [comment]
So what about this:

- create a `release` branch that has our major releases and tags
- generate `minimumSchema.graphql`
- *no* breaking changes to `master` ever!
- we would add `@deprecation` directives on all fields intended to change or be removed
- when we're ready to release, remove deprecation and merge `master` into `release`
- the `release` branch does not check against `minimumSchema.graphql`
- publish `release` to npm

**Comment by vinnyjth** [comment]
@redreceipt I think that's a great plan when we have more confidence around when and where we would allow breaking changes, and when we have more confidence around our future plans with branches. 

In the meantime, I would prefer we treat the remote production API as the version of the scheme to check against. You can use something like this to download a remote schema
https://github.com/prisma/get-graphql-schema

**Comment by redreceipt** [comment]
Word, I like that. So until then, we just won't allow any breaking changes anywhere.

**Comment by redreceipt** [comment]
@vinnyjth so here's a demo of using option 2, by changing this:

<img width="378" alt="Screen Shot 2019-05-31 at 2 56 51 PM" src="https://user-images.githubusercontent.com/2659478/58728199-66967880-83b4-11e9-8eae-9b577d588f7f.png">

We get this:

```
~/Documents/Projects/apollos-prototype/packages/apollos-church-api (master *%) $ npx @entria/graphql-findbreakingchanges old.graphql schema.graphql 
BREAKING CHANGES
{ type: 'FIELD_REMOVED',
  description: 'ThemeColors.alert was removed.' }
```

Seems to have everything we would need.

What would you say would be the best method for updating the local schema? We could use a post-commit hook?

**Comment by vinnyjth** [comment]
@redreceipt I think the best option is generating the schema on the fly in the test environment. One way to do that is to start the server, use the tool above to generate the schema, and then stop the server.

**Comment by redreceipt** [comment]
O gotcha. Well, we'd need to keep a local schema file if we want to lint, which I think would be a major plus.

**Comment by vinnyjth** [comment]
Oh right! Well I would do it as a `postinstall` step. Look at how the `apolloschurchapp` generates the `fragmentTypes.json` file.
